docs/posts/llms-and-data/index.qmd

```diff
@@ -0,0 +1,258 @@
+
+
+@marvin.ai_model  # <3>
+class VowelCounter(BaseModel):  # <3>
+    """Count vowels in a string."""  # <3>
+
+    include_y: bool = Field(False, description="Include 'y' as a vowel.")  # <3>
+    # num_a: int = Field(..., description="The number of 'a' vowels.") # <3>
+    # num_e: int = Field(..., description="The number of 'e' vowels.") # <3>
+    # num_i: int = Field(..., description="The number of 'i' vowels.") # <3>
+    # num_o: int = Field(..., description="The number of 'o' vowels.") # <3>
+    # num_u: int = Field(..., description="The number of 'u' vowels.") # <3>
+    # num_y: int = Field(..., description="The number of 'y' vowels.") # <3>
+    num_total: int = Field(..., description="The total number of vowels.")  # <3>
+
+
+VowelCounter("hello world")  # <4>
+```
+
+1. Additional imports for Pydantic.
+2. Configure Marvin to use a cheaper model.
+3. A non-deterministic, LLM-powered AI model.
+4. Call the AI model on some text.
+
+Then we'll have the LLM write the UDF, just to be fancy:
+
+```{python}
+udf = create_udf_from_text(
+    "a function named count_vowels_ai that given an input string, calls VowelCounter on it and returns the num_total attribute of that result"
+)
+print(udf)
+exec(udf)
+```
+
+```{python}
+t5 = t3.mutate(
+    species_vowel_count=count_vowels_ai(t3.species),
+    island_vowel_count=count_vowels_ai(t3.island),
+)
+t5
+```
+
+Notice that in this UDF, unlike in the previous example, a LLM is being
+called (possibly several times) for each row in the table. This is a very expensive
+operation and we'll need to be careful about how we use it in practice.
+
+## Summary
+
+To summarize this post:
+
+```{python}
+from rich import print
+
+with open("index.qmd", "r") as f:
+    self_text = f.read()
+
+
+@marvin.ai_model
+class Summary(BaseModel):
+    """Summary of text."""
+
+    summary_line: str = Field(..., description="The one-line summary of the text.")
+    summary_paragraph: str = Field(
+        ..., description="The one-paragraph summary of the text."
+    )
+    conclusion: str = Field(
+        ..., description="The conclusion the reader should draw from the text."
+    )
+    key_points: list[str] = Field(..., description="The key points of the text.")
+    critiques: list[str] = Field(
+        ..., description="Professional, fair critiques of the text."
+    )
+    suggested_improvements: list[str] = Field(
+        ..., description="Suggested improvements for the text."
+    )
+    sentiment: float = Field(..., description="The sentiment of the text.")
+    sentiment_label: str = Field(..., description="The sentiment label of the text.")
+    author_bias: str = Field(..., description="The author bias of the text.")
+
+
+print(Summary(self_text))
+```
+
+## Next steps
+
+You can get involved with [Ibis
+Birdbrain](https://github.com/ibis-project/ibis-birdbrain), our open-source data
+& AI project for building next-generation natural language interfaces to data.
+
+## Discussions
```

Comment:
Is this empty `## Discussions` section intentional?

Reply:
yeah I'm going to remove it though -- it's before the Giscus discussions at the bottom, but would look weird via RSS and it's obvious w/o a header